ext/opcache/zend_accelerator_module.c

```diff
@@ -363,6 +365,37 @@ static int filename_is_in_cache(zend_string *filename)
 	return 0;
 }
 
+static int filename_is_in_file_cache(zend_string *filename)
+{
+	if (!ZCG(accel_directives).file_cache) {
+		return 0;
+	}
+
+	zend_string *realpath;
+
+	realpath = zend_resolve_path(filename);
+
+	if (!realpath) {
+		return 0;
+	}
+
+	zend_file_handle handle;
+
+	zend_stream_init_filename_ex(&handle, filename);
+	handle.opened_path = realpath;
+
+	zend_persistent_script *persistent_script = zend_file_cache_script_load(&handle, true);
```

Comment:
Will this lead to loading from the file cache?
This look weird.

Reply:
Kind of, the `, true` parameter was to skip loading into SHM, so it'll effectively fully load the file & unserialize it, just not load it into SHM.

I'll agree this probably wasn't ideal.

How about this latest commit, swapping to `zend_file_cache_script_validate(...)` ?

I was unhappy with the code re-use, as ideally I'd want to replace the validation code within `zend_file_cache_script_load(...)` with a call to `zend_file_cache_script_validate(...)`:

However, this had the side effect that I'd either be opening, reading & locking the cache file twice, or, we'd have to return a struct from `zend_file_cache_script_validate(...)` - something like this:

```c
// New struct to hold validation results and the file descriptor
typedef struct _zend_file_cache_validation_result {
    zend_file_cache_metainfo info;
    int fd;
    off_t file_size;
} zend_file_cache_validation_result;


// Modified validation function to return the file descriptor and file size
static zend_file_cache_validation_result *zend_file_cache_script_validate(zend_file_handle *file_handle)
{
    ...
}
```

and that felt even more complicated.... So I'm at a loss really.

How would you handle it?

And if we go with a new function as @iluuu1994 has suggested, does anyone have a recommendation for the function name?

Reply:
Adding a parameter to `zend_file_cache_script_load()` would be the simplest solution, but a separate function as you did is a better API (IMHO).

I like your suggestion of making the validate function return a struct, but I suggest to pass the struct as parameter so that the caller can allocate it on the stack:

``` c
zend_result zend_file_cache_script_validate(zend_file_cache_validation_result *result);
```

Maybe rename `zend_file_cache_script_validate()` to `zend_file_cache_open()`, and `zend_file_cache_validation_result` to `zend_file_cache_handle`, and add a small wrapper of `zend_file_cache_open()` that doesn't take this parameter:

``` c
zend_always_inline zend_result zend_file_cache_open(zend_file_handle *file_handle, zend_file_cache_handle *cache_handle);

zend_result zend_file_cache_validate(zend_file_handle *file_handle)
{
    zend_file_cache_handle cache_handle;
    return zend_file_cache_open(file_handle, &cache_handle);
}

// maybe also:
zend_always_inline zend_file_cache_close(zend_file_cache_handle *cache_handle);
```

Reply:
Thanks @arnaud-lb ,

I've just pushed my first try at implementing the way you've suggested.

I actually really like that approach and it makes a lot more logical sense.

In the end, rather than passing around a file descriptor, since we were already reading the whole file to potentially checksum it, I've opted to pass around the `zend_arena_checkpoint` and `zend_arena_alloc` values:

I won't pretend I understand this very well, so if I've done something wrong, please speak up!

Reply:
Sorry, I overlooked the arena usage before my previous comment. Unfortunately, passing around the arena checkpoint would result in fragile code.

Alternatives would be to allocate the buffer in some other way, or to mmap the file, but this is too much changes and they may negatively impact the performance of `zend_file_cache_script_load()`.

At this point I think that adding a `validate_only` parameter to `zend_file_cache_script_load()` is a better solution, after all.

Reply:
No worries, getting your originally suggested method to build & work has been a great learning experience 😃 

Thanks for your time reviewing and helping to move this along, it's really appreciated. 

> At this point I think that adding a `validate_only` parameter to `zend_file_cache_script_load()` is a better solution, after all.

This is pretty much what I'd done in my first attempt that @dstogov didn't seem to like, so before I switch back:

@dstogov are you happy with that?